### PR TITLE
Enabled shimming of external applications

### PR DESCRIPTION
**Pull request by excitoon**
Hi.

Now one can make shims for external applications (system, other applications installed by scoop and so on).

First, `$bin` is to be searched in `$dir`, then globally, then in user and system `PATH`.

Examples:
```
    "bin": [ [ "winver.exe", "myver", "" ] ]
```

```
    "depends": [ "python" ],
    "bin": [
        [ "python.exe", "myscript", "$dir/myscript.py" ],
        [ "python.exe", "myscript.py", "$dir/myscript.py" ]
    ]
```

```
    "depends": [ "bash" ],
    "bin": [
        [ "bash.exe", "myscript", "$dir/myscript.sh" ],
        [ "bash.exe", "myscript.sh", "$dir/myscript.sh" ]
    ]
```

### CODE REVIEW

**Comment by r15ch13** [comment]
Do you have an example for which this could be useful?

**Comment by excitoon** [comment]
It could be useful for virtually any case where script is being shipped without interpreter. For example, `openssh` ships with few Bash scripts (`findssl.sh` and `ssh-copy-id`) and it forces maintainer to include `bash` and its dependencies into manifest (and to file system of every user).
More examples (Python, Perl, Bash):
- getmail
- SimpleHTTPServer (one would shim it without any real files, just Python 2)
- deluge
- newspipe
- [https://github.com/jpoliv/wakeonlan](wakeonlan)
- [https://github.com/jpoliv/wakeonlan](swaddle)

**Comment by excitoon** [comment]
Basically my point is that every programmer who writes scripts and want to ship it to Windows systems need to make some `cmd` wrapper around them in order to run them from command line. This patch will allow run these scripts in Windows in the same way as in Unix.

**Comment by gbursson** [comment]
excitoon, just to let you know, two links (for wakeonlan and swaddle) are incorrect

**Comment by stkb** [comment]
I've also found this patch useful for a tool I've been using (which is actually a php script).

File: lib/install.ps1 [hunk 1]
```diff
@@ -679,14 +679,16 @@ function create_shims($manifest, $dir, $global, $arch) {
         $target, $name, $arg = shim_def $_
         write-output "Creating shim for '$name'."
 
-        # check valid bin
-        $bin = "$dir\$target"
-        if(!(is_in_dir $dir $bin)) {
-            abort "Error in manifest: bin '$target' is outside the app directory."
+        if(test-path "$dir\$target") {
```
Comment (stkb) [inline lib/install.ps1]: `test-path` should probably have `-pathType leaf` to be safe

File: lib/install.ps1 [hunk 2]
```diff
@@ -679,14 +679,16 @@ function create_shims($manifest, $dir, $global, $arch) {
...
+            $bin = "$dir\$target"
+        } elseif(test-path $target) {
```
Comment (stkb) [inline lib/install.ps1]: Not sure what this check is for (if $target is absolute?)

File: lib/install.ps1 [hunk 3]
```diff
@@ -679,14 +679,16 @@ function create_shims($manifest, $dir, $global, $arch) {
...
+            $bin = $target
+        } else {
+            $bin = search_in_path $target
         }
         if(!(test-path $bin)) { abort "Can't shim '$target': File doesn't exist."}
```
Comment (stkb) [inline lib/install.ps1]: This needs to be something like `if(!$bin) { ... }` instead. Currently, if we don't find the file, `$bin` is null and this check blows up, causing lots of errors as the null value propogates. 
Also there's no need to do the `test-path` again.

File: lib/install.ps1 (hunk 2 above)
Comment (excitoon) [reply lib/install.ps1]: Yeah, just because absolute path is better than invalid path.

File: lib/install.ps1 (hunk 3 above)
Comment (excitoon) [reply lib/install.ps1]: Yeah, you're right.

File: lib/install.ps1 (hunk 1 above)
Comment (excitoon) [reply lib/install.ps1]: Thanks. Actually I did not use Powershell before.

**Comment by excitoon** [comment]
Rebased first commit on current master.

**Comment by excitoon** [comment]
One more thing about `For-Each` and piping:
```
    $path.split(';') | % {
        if(test-path "$_\$target" -pathType leaf) {
            return "$_\$target"
        }
    }
```

If you write it like that, function may return a list consisting of joined return values from each element. That's totally counter-intuitive. If one replace that with regular `foreach` everything works as expected.

**Comment by excitoon** [comment]
I guess it is ready to merge/review again. Tested on all mentioned cases.

**Comment by excitoon** [comment]
Hi @r15ch13. Can you merge it (and the other 3 pull requests: #2079, #2118, #2120)?

**Pull request merged**